feat: add and use getBlobsV1 to expedite gossip import

g11tech opened the pull request:
one may review but still

- [x] undergoing testing/observation
![image](https://github.com/user-attachments/assets/6734cacf-9fdf-4070-b11f-6fdc62ff3590)


Closes #7114

pending todo for a followup PR
- add dashboard panels for the metrics

twoeths commented on packages/beacon-node/src/chain/blocks/verifyBlocksDataAvailability.ts:
```diff
@@ -43,9 +44,12 @@ export async function verifyBlocksDataAvailability(
 
   for (const blockInput of blocks) {
+    if (signal.aborted) {
+      throw new ErrorAborted("verifyBlockStateTransitionOnly");
```
```suggestion
      throw new ErrorAborted("verifyBlocksDataAvailability");
```

twoeths commented on packages/beacon-node/src/chain/blocks/verifyBlocksDataAvailability.ts:
```diff
@@ -122,16 +127,21 @@ async function maybeValidateBlobs(
+  const cutoffTimeout =
+    cutoffTime > 0
+      ? new Promise((_resolve, reject) => {
+          setTimeout(reject, cutoffTime);
```
we should add reason to the reject functions, otherwise we'll get "undefined" reason which is very hard to debug

twoeths commented on packages/beacon-node/src/network/reqresp/beaconBlocksMaybeBlobsByRoot.ts:
```diff
@@ -77,26 +82,62 @@ export async function unavailableBeaconBlobsByRoot(
-  if (blobIdentifiers.length > 0) {
-    allBlobSidecars = await network.sendBlobSidecarsByRoot(peerId, blobIdentifiers);
+  const networkReqIdentifiers: deneb.BlobIdentifier[] = [];
+  if (executionEngine !== undefined) {
```
`IBeaconChain` always has `executionEngine` so it should not be an optional param

twoeths commented on packages/beacon-node/src/network/reqresp/beaconBlocksMaybeBlobsByRoot.ts:
```diff
@@ -77,26 +82,62 @@ export async function unavailableBeaconBlobsByRoot(
+  if (executionEngine !== undefined) {
+    const signedBlockHeader = signedBlockToSignedHeader(config, block);
+    const versionedHashes = blobIdentifiersWithCommitments.map((bi) => kzgCommitmentToVersionedHash(bi.kzgCommitment));
+    const blobAndProofs = await executionEngine
```
`executionEngine.getBlobs()` should not be called inside this function because the consumer of this function (which is `unknownBlock`) has a loop over shuffled n peers, it would means that this function is called n times

g11tech replied:
we can keep a flag that it has been called before but the main problem is that when we don't even know the block, we can't call this since we dont know versioned hashes, and we need to call this before start req/resp of blobs from the network

so the best way would be to just track it via  a flag in pending blockinput

g11tech replied:
actually have locally modified it to use a pruneable fixed length cache, that gets populated whenever there is a "new" versioned hash to be queried (same tx in reorgs etc), testing that out

nflaig commented on packages/beacon-node/src/execution/engine/http.ts:
```diff
@@ -463,6 +465,17 @@ export class ExecutionEngineHttp implements IExecutionEngine {
     return response.map(deserializeExecutionPayloadBody);
   }
 
+  async getBlobs(_fork: ForkName, versionedHashes: VersionedHashes): Promise<(BlobAndProof | null)[]> {
```
if the EL does not have this api implemented yet it will always result in an error, from my observation when we implemented the client version api, this can lead to engine status switching from syncing to offline

nflaig commented on packages/beacon-node/src/execution/engine/http.ts:
```diff
@@ -463,6 +465,17 @@ export class ExecutionEngineHttp implements IExecutionEngine {
+    const method = "engine_getBlobsV1";
+    assertReqSizeLimit(versionedHashes.length, 128);
+    const versionedHashesHex = versionedHashes.map(bytesToData);
+    const response = await this.rpc.fetchWithRetries<
```
we should probably stop trying to query this api if it failed earlier with specific code that method is not implemented and just return `null`

and since we wanna re-check if the EL is updated, could reset that behavior and retry querying if EL becomes online
https://github.com/ChainSafe/lodestar/blob/068fbae928989295706443887c645d476bb95695/packages/beacon-node/src/execution/engine/http.ts#L496-L501

g11tech replied:
added a mechanism to not query for a while if this fails with method not found

wemeetagain commented on packages/beacon-node/src/chain/blocks/verifyBlocksDataAvailability.ts:
```diff
@@ -122,16 +127,21 @@ async function maybeValidateBlobs(
-  );
-  const cutoffTimeout = new Promise((_resolve, reject) => setTimeout(reject, cutoffTime));
+  const cutoffTime =
+    computeTimeAtSlot(chain.config, blockSlot, chain.genesisTime) * 1000 + BLOB_AVAILABILITY_TIMEOUT - Date.now();
```
why was the < 0 case changed? Before, when cutoffTime was < 0, it was set to 0, now it rejects.

g11tech replied:
should still resolve the race properly as dscussed

wemeetagain commented on packages/beacon-node/src/execution/engine/http.ts:
```diff
@@ -463,6 +467,56 @@ export class ExecutionEngineHttp implements IExecutionEngine {
+        throw e;
+      });
+
+    // handle nethermind buggy response
```
is there an issue we can link to here?

g11tech replied:
not really as issue, it was just a debugging conversation with 1.29.1 having this invalid response (We can clean it up a bit later when people would have moved off 1.29.1)

https://discord.com/channels/595666850260713488/1293605631785304088/1298956894274060301

wemeetagain commented on packages/beacon-node/src/execution/engine/mock.ts:
```diff
@@ -396,6 +397,12 @@ export class ExecutionEngineMockBackend implements JsonRpcBackend {
   }
 
+  private getBlobs(
+    versionedHashes: EngineApiRpcParamTypes["engine_newPayloadV3"][1]
```
can we add a RpcParamTypes entry for this endpoint?

wemeetagain commented on packages/beacon-node/src/metrics/metrics/beacon.ts:
```diff
@@ -205,6 +205,121 @@ export function createBeaconMetrics(register: RegistryMetricCreator) {
 
+    blockInputFetchStats: {
+      // of already available blocks which didn't have to go through blobs pull
+      totalDataAvailableBlockInputs: register.gauge({
```
I think these can mostly be counters

g11tech replied:
we don't use counters yet

g11tech committed:
hookup the getblobs api to get bob and proof data from el

remove unused

fix import

metrics overhault, test, debugging testing, some feeback

fix

add nethermind bug dicussion link

fix

resolve conflicts

g11tech committed:
deblobs timeout

g11tech committed:
fix metric

twoeths commented:
mesh peers on the unstable mainnet node dropped consistently since this PR merged

<img width="1514" alt="Screenshot 2024-11-06 at 10 17 05" src="https://github.com/user-attachments/assets/626e29ed-3867-4fcc-997f-594b58a8230f">

same to `blob_sidecar` topic, it does not happen on holesky through

wemeetagain commented:
:tada: This PR is included in [v1.24.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.24.0) :tada:
